soci/store/store.go

```diff
@@ -0,0 +1,356 @@
+
+	// Default path to soci content addressable storage
+	DefaultSociContentStorePath = "/var/lib/soci-snapshotter-grpc/content"
+)
+
+type Config struct {
+	Type      ContentStoreType `toml:"type"`
+	Namespace string           `toml:"namespace"`
+}
+
+func NewStoreConfig(opts ...Option) Config {
+	storeConfig := Config{
+		Type:      DefaultContentStoreType,
+		Namespace: namespaces.Default,
+	}
+	for _, o := range opts {
+		o(&storeConfig)
+	}
+	return storeConfig
+}
+
+type Option func(*Config)
+
+func WithNamespace(namespace string) Option {
+	return func(sc *Config) {
+		sc.Namespace = namespace
+	}
+}
+
+func WithType(contentStoreType ContentStoreType) Option {
+	return func(sc *Config) {
+		sc.Type = contentStoreType
+	}
+}
+
+func ErrUnknownContentStoreType(contentStoreType ContentStoreType) error {
+	return fmt.Errorf("unknown content store type: %s; must be one of %s or %s",
+		contentStoreType, ContainerdContentStoreType, SociContentStoreType)
+}
+
+// CanonicalizeContentStoreType resolves the empty string to DefaultContentStoreType,
+// returns other types, or errors on unrecognized types.
+func CanonicalizeContentStoreType(contentStoreType ContentStoreType) (ContentStoreType, error) {
+	switch contentStoreType {
+	case "":
+		return DefaultContentStoreType, nil
+	case ContainerdContentStoreType, SociContentStoreType:
+		return contentStoreType, nil
+	default:
+		return "", ErrUnknownContentStoreType(contentStoreType)
+	}
+}
+
+// GetContentStorePath returns the top level directory for the content store.
+func GetContentStorePath(contentStoreType ContentStoreType) (string, error) {
+	contentStoreType, err := CanonicalizeContentStoreType(contentStoreType)
+	if err != nil {
+		return "", err
+	}
+	switch contentStoreType {
+	case ContainerdContentStoreType:
+		return DefaultContainerdContentStorePath, nil
+	case SociContentStoreType:
+		return DefaultSociContentStorePath, nil
+	default:
+		return "", ErrUnknownContentStoreType(contentStoreType)
+	}
+}
+
+// GetContentStoreBlobPath returns the bottom level directory for the content store, e.g. "/blobs/sha256".
+func GetContentStoreBlobPath(contentStoreType ContentStoreType) (string, error) {
+	contentStorePath, err := GetContentStorePath(contentStoreType)
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(contentStorePath, "blobs", "sha256"), nil
+}
+
+type CleanupFunc func(context.Context) error
+
+func nopCleanup(context.Context) error { return nil }
+
+func NewContentStore(ctx context.Context, opts ...Option) (context.Context, Store, error) {
+	storeConfig := NewStoreConfig(opts...)
+
+	contentStoreType, err := CanonicalizeContentStoreType(storeConfig.Type)
+	if err != nil {
+		return ctx, nil, err
+	}
+	switch contentStoreType {
+	case ContainerdContentStoreType:
+		return NewContainerdStore(ctx, storeConfig)
+	case SociContentStoreType:
+		return NewSociStore(ctx)
+	default:
+		return ctx, nil, ErrUnknownContentStoreType(contentStoreType)
```

Comment:
nit: Isn't this check done above in `CanonicalizeContentStoreType`?

Reply:
Yes and it's omitted everywhere else that CanonicalizeContentStoreType gets used, but right here I need a default return either in the case or after the case to satisfy the compiler even though we know execution will never reach it.

Reply:
I see. Maybe we can just `break`?

Reply:
I've made the unreachable return more unique and explicit.